### PR TITLE
Prometheus metrics for L7 proxy requests.

### PR DESCRIPTION
**Pull request by manalibhutiyani**
Covers adding the following:

- Rejected L7 (HTTP/Kafka) due to parse errors.
- Forwarded L7 (HTTP/Kafka)
- Denied L7 (HTTP/Kafka) due to policy

 Fixes: #4010
 Related to: #3337
    
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

```release-note
Add prometheus metrics for L7 proxy requests
```

### CODE REVIEW

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by joestringer** [comment]
Does the URL fix need backporting? Should it be a separate PR for that reason?

**Comment by manalibhutiyani** [comment]
Sure I can separate it out in a different PR.

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
@joestringer Opened a separate PR #4039 for the docs change.

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -63,6 +63,7 @@ import (
 	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/rest"
 
+	"github.com/cilium/cilium/pkg/metrics"
```
Comment (rlenglet) [inline pkg/endpoint/endpoint.go]: Move this import in the Cilium imports group above.

File: pkg/endpoint/endpoint.go [hunk 2]
```diff
@@ -2093,14 +2094,18 @@ func (e *Endpoint) UpdateProxyStatistics(l7Protocol string, port uint16, ingress
...
+	metrics.ProxyReceived.Inc()
 
 	switch verdict {
 	case accesslog.VerdictForwarded:
 		stats.Forwarded++
+		metrics.ProxyForwarded.Inc()
 	case accesslog.VerdictDenied:
 		stats.Denied++
+		metrics.ProxyDenied.Inc()
 	case accesslog.VerdictError:
```
Comment (rlenglet) [inline pkg/endpoint/endpoint.go:2106]: Note that this code branch is dead. Kafka is never emitting log messages with this verdict. That needs to be fixed in the Kafka proxy.

Comment (rlenglet) [reply pkg/endpoint/endpoint.go:2106]: Please add unit tests for all types of verdicts to check that the metrics get incremented as expected.

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -129,6 +129,35 @@ var (
 		Help:        "Last timestamp when we received an event",
 		ConstLabels: prometheus.Labels{"source": LabelEventSourceAPI},
 	})
+
+	// Proxy statistics
+	// ProxyParseErrors is a count of failed parse errors on proxy
+	ProxyParseErrors = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "proxy_parse_errors",
```
Comment (rlenglet) [inline pkg/metrics/metrics.go]: Please follow Prometheus's naming conventions:
https://prometheus.io/docs/practices/naming/
For instance, I'd suggest naming those metrics:
```
proxy_parse_errors_total
proxy_forwarded_requests_total
proxy_denied_requests_total
proxy_received_requests_total
```

**Review by rlenglet — DISMISSED** [review]

File: pkg/endpoint/endpoint.go (hunk 2 above)
Comment (manalibhutiyani) [reply pkg/endpoint/endpoint.go:2106]: @rlenglet Raised #4041 to fix the Kafka parser. I will do that as a separate PR. Thanks!

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (manalibhutiyani) [reply pkg/metrics/metrics.go]: done

File: pkg/endpoint/endpoint.go (hunk 1 above)
Comment (manalibhutiyani) [reply pkg/endpoint/endpoint.go]: done

File: pkg/endpoint/endpoint.go (hunk 2 above)
Comment (manalibhutiyani) [reply pkg/endpoint/endpoint.go:2106]: @rlenglet raised #4043 for adding the CI tests as there is not a non-generic way of unit testing this. Thanks.

**Comment by manalibhutiyani** [comment]
@rlenglet Addressed review comments and raised GH issues for fixing Kafka parser errors and CI tests. Please take a look.

**Review by rlenglet — APPROVED** [review]

**Comment by manalibhutiyani** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

File: pkg/endpoint/endpoint.go [hunk 4]
```diff
@@ -2093,14 +2094,18 @@ func (e *Endpoint) UpdateProxyStatistics(l7Protocol string, port uint16, ingress
 	}
 
 	stats.Received++
+	metrics.ProxyReceived.Inc()
 
 	switch verdict {
 	case accesslog.VerdictForwarded:
 		stats.Forwarded++
+		metrics.ProxyForwarded.Inc()
```
Comment (tgraf) [inline pkg/endpoint/endpoint.go:2102]: How expensive is this operation? Do we need to update a local counter and sync to the metrics periodically or is the Prometheus client side code smart enough to not perform a heavy increment?

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -129,6 +129,36 @@ var (
 		Help:        "Last timestamp when we received an event",
 		ConstLabels: prometheus.Labels{"source": LabelEventSourceAPI},
 	})
+
+	// Proxy statistics
+
+	// ProxyParseErrors is a count of failed parse errors on proxy
+	ProxyParseErrors = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "proxy_parse_errors_total",
```
Comment (tgraf) [inline pkg/metrics/metrics.go]: I recommend to not call these proxy_. Users will not know what it means. We should also make sure that it is clear that these are relevant when policy enforcement is active.

policy_l7_parse_error_total
policy_l7_forwarded_total
policy_l7_denied_total
policy_l7_processed_total

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by tgraf** [comment]
Please add the metrics to:
http://docs.cilium.io/en/doc-1.0/configuration/metrics/#exported-metrics

This also ensures that we maintain a stable set of prefixes for all metrics

File: pkg/endpoint/endpoint.go (hunk 4 above)
Comment (manalibhutiyani) [reply pkg/endpoint/endpoint.go]: I am not sure of this one. @raybejjani might be able to answer this better in terms of cost. I followed the convention we use for existing prometheus metrics.

File: pkg/metrics/metrics.go (hunk 5 above)
Comment (manalibhutiyani) [reply pkg/metrics/metrics.go]: sure.

File: pkg/endpoint/endpoint.go (hunk 4 above)
Comment (manalibhutiyani) [reply pkg/endpoint/endpoint.go:2102]: It looks like we update a local counter atomically using this code:
```
func (c *counter) Inc() {
	atomic.AddUint64(&c.valInt, 1)
}
```
Trying to see how we sync this information with the server side with these metrics.

Comment (raybejjani) [reply pkg/endpoint/endpoint.go:2102]: The collection frequency is configured by the prometheus collector (we use 10s in our examples). I expect each collection request ends up in
https://github.com/prometheus/client_golang/blob/master/prometheus/counter.go#L110-L116
```golang
func (c *counter) Write(out *dto.Metric) error {
	fval := math.Float64frombits(atomic.LoadUint64(&c.valBits))
	ival := atomic.LoadUint64(&c.valInt)
	val := fval + float64(ival)

	return populateMetric(CounterValue, val, c.labelPairs, out)
}
```
for each request. In any case, though, the cost of these reads is in the collection request, not during the `Inc` we do in the fast path, so this shouldn't cause any slowdown we would worry about.

File: pkg/metrics/metrics.go (hunk 5 above)
Comment (manalibhutiyani) [reply pkg/metrics/metrics.go]: done

**Comment by manalibhutiyani** [comment]
@tgraf : Addressed review comments. Please take a look.

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
test-docs-please

File: pkg/endpoint/endpoint.go (hunk 4 above)
Comment (tgraf) [reply pkg/endpoint/endpoint.go:2102]: 👍

**Review by tgraf — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -129,6 +129,36 @@ var (
...
+		Name:      "policy_l7_forwarded_total",
+		Help:      "Number of total L7 forwarded requests/responses",
+	})
+
+	// ProxyDenied is a count of all denied requests by policy by the proxy
+	ProxyDenied = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "policy_l7_denied_total",
+		Help:      "Number of total L7 denied requests/responses due to policy",
+	})
```
Comment (raybejjani) [inline pkg/metrics/metrics.go:154]: I'm on the fence about separate `policy_l7_denied_total` and  `policy_l7_forwarded_total`. It doesn't really matter, since one can sum these two to get the total number of L7 requests. On the other hand, I would have expected the policy_l7_total to give me the total, and I could then split them out by label (`forwarded` and `denied`).
Similarly, what is the relationship with `policy_l7_recieved_total` below, and `policy_l7_parse_errors_total` above? Are some of these egress and others ingress? Those would be candidates for labels as well/instead.
This is mostly based on my reading of https://prometheus.io/docs/practices/naming/, especially the HTTP examples.

**Review by raybejjani — APPROVED** [review]
Can you also run the metric name linter https://manpages.debian.org/testing/prometheus/promtool.1.en.html#check_metrics before merging?

I don't want to block this, so it is approved, but I would prefer us to figure out if we meet prometheus style, especially as we add a lot of new metrics. This was raised at the community meeting week. My concern is that tooling around prometheus would expect one thing (aggregate by label) and we would do another (aggregate by metric name) and that this would make the metrics difficult to use.

File: pkg/metrics/metrics.go (hunk 6 above)
Comment (manalibhutiyani) [reply pkg/metrics/metrics.go:154]: We already collect these separate stats for Proxy and I simply exported them over. My understanding is we need them all even if we can derive one from the sum/total, the fact that the sums might not add up should give some clue as to if something is going wrong? This will help understand if cilium is behaving incorrectly, so I exported all the metrics we collect by default in proxy.
Let me know your thoughts. Since this is more of a first step, we can always iterate and make this better.

These metrics are the sum total of both ingress/egress (see `UpdateProxyStatistics`) and are simply calculating the total of the requests/response they see.  So `policy_l7_forwarded_total` should be 2 for a legit HTTP request and corresponding response.
The GH issue did not ask for ingress/egress separation, so as a beginning step I simply exported the values we already have in the Proxy. 

`policy_l7_recieved_total` are the total number of packets received independent of whether they were denied, dropped et al.

`policy_l7_parse_errors_total` are due to errors during parsing of the requests/responses (not necessarily due to policy denial, but more like a corrupt request / something our parser does not support et al)

Comment (manalibhutiyani) [reply pkg/metrics/metrics.go:154]: @raybejjani : I actually found issues with the metrics already present. Promtool did not complain about any of the L7 proxy metrics I have added as part of this PR.

```
/prometheus # cat metrics.txt  | promtool check-metrics
cilium_endpoint_regenerations: counter metrics should have "_total" suffix
cilium_policy_import_errors: counter metrics should have "_total" suffix
cilium_endpoint_count: non-histogram and non-summary metrics should not have "_count" suffix
cilium_policy_count: non-histogram and non-summary metrics should not have "_count" suffix
/prometheus # 
```

I have raised GH #4053  to address this separately. Thanks!

Comment (raybejjani) [reply pkg/metrics/metrics.go:154]: Yeah, the already present ones are bad, but they're part of the API now and I'm not sure how stable we want to keep the metrics API. In that sense, by merging this PR we are also committing to this naming scheme for these new ones.

> We already collect these separate stats for Proxy and I simply exported them over.

It's not that relevant what the proxy tracks, though. These metrics are not the internal cilium metrics, and we shouldn't assume they are (although it would be useful to export those in some debug section). These are operational metrics that an operator would use to characterise the health of the cluster or track traffic/usage patterns. In that sense, these aren't the internal state exactly, but the user-relevant equivalent. An example would be: endpoints have an internal state machine but every state may not interest the user, they might only care about `ready`, `not-ready`, `regenerating`, so we would only emit metrics with those labels (note that `not-ready` isn't even a state!).
My point was that it seems like there is one total: `policy_l7_recieved_total`. Of this total, some of the increments represent forwarded requests, some denied requests, or whatever else happens to these requests. I'm assuming these actions are exclusive. If they aren't then this applies less (we should update the descriptions to make that more clear if so). My impression is that the prevailing trend is to use labels on metrics to distinguish the states of the things being counted or measured e.g. you have one http_requests_total and each response code is a label. This allows a more top-down view into the metric: first you have a simple total, then you begin slicing it with label selectors e.g. you have http_request_duration_seconds or http_request_total, then averages, min, max, percentiles represent all http traffic. If you care about only 200 requests, you select for that only. If you want to exclude 404s because they skew the statistics, you then exclude those.
This isn't to say that multiple metrics wouldn't work in these scenarios. Older schemes did exactly that (but it's possible newer metrics systems make it hard to combine separate metrics this way, I'm not sure). I think they fell out of favour because they disallowed mixing multiple labels e.g. the http response code and also a QoS label per request. In our case, the equivalent would be how a user of these metrics wants to combine or split up this count. Are `policy_l7_forwarded_total`, `policy_l7_denied_total` and `policy_l7_received_total` always semantically different? or do they represent the same things in different ways? When I come to build a dashboard, what do I want to do with each of these independent metrics?

err, this came out as more of a rant. It was more an exercise for me to lay out my thinking.
tl;dr is that the metrics are an interface and we probably need to keep it stable: users will build alerts on them and we can't break the alerts.

Comment (manalibhutiyani) [reply pkg/metrics/metrics.go:154]: `These metrics are not the internal cilium metrics, and we shouldn't assume they are (although it would be useful to export those in some debug section)`

Actually they are. They track the parse errors /request denies / request forwarded by different proxies we use (Envoy / Kafka golang proxy).  I am not sure what you mean by `user-relevant equivalent` but the GH issue  #3337 and #4010 specially asked for these specific metrics:
```
Rejected L7 (HTTP/Kafka) due to parse errors.
Forwarded L7 (HTTP/Kafka)
Denied L7 (HTTP/Kafka) due to policy
```
And these are the exact metrics that are being exposed in this PR.
I think there is some confusion about why these are not internal cilium metrics.
Also, there is a subtle difference between `policy_l7_forwarded_total` , `policy_l7_denied_total` and `policy_l7_received_total`. We can obviously derive one from other, but I think the main purpose(use-case which Amey is working on) was to find out if cilium is misbehaving in  any of the data paths. Again maybe this does not fall under standard prometheus usage and maybe we can discuss your concerns offline and I can be more mindful about your concerns next time. Also, I do agree that we can aggregate these with the help of relevant labels, but my initial rationale was that these 3 are different metrics and should be exposed nevertheless.

Comment (raybejjani) [reply pkg/metrics/metrics.go:154]: You're correct. I should have commented on the PR when it was first made instead of the PR. Sorry.
